src/spatial/utils.jl

```diff
@@ -26,191 +22,358 @@ abstract type AbstractSpatialSystem end
-struct SpatialRates{R} <: AbstractSpatialRates
-    rx_rates::Matrix{R} # rx_rates[i,j] is rate of reaction i at site j
-    hop_rates::Matrix{R} # hop_rates[i,j] is rate of species i at site j
-    rx_rates_sum::Vector{R} # [sum of reaction rates for site 1,...,sum of reaction rates for last site]
-    hop_rates_sum::Vector{R} # [sum of hopping rates for site 1,...,sum of hopping rates for last site]
+function update_hop_rate!(hop_rates::HopRatesUnifNbr, species::Int, u, site, spatial_system) 
+    set_hop_rate_at_site!(hop_rates, site, species, evalhoprate(hop_rates, u, species, site, num_neighbors(spatial_system, site)))
 end
 
+# helper functions
 """
-standard constructor, assumes numeric values for rates
+returns hopping rates of a site
 """
-function SpatialRates(rx_rates::Matrix{R}, hop_rates::Matrix{R}) where {R}
-    num_sites = size(rx_rates, 2)
-    SpatialRates{R}(rx_rates, hop_rates, [sum(@view rx_rates[:,i]) for i in 1:num_sites], [sum(@view hop_rates[:,i]) for i in 1:num_sites])
+function hop_rates_at_site(hop_rates::HopRatesUnifNbr, site) 
+    @view hop_rates.rates[:,site]
 end
 
 """
-initializes SpatialRates with zero rates
+sets the rate of hopping at site. Return the old rate
 """
-function SpatialRates(numrxjumps::Integer,num_species::Integer,num_sites::Integer)
-    reaction_rates = zeros(Float64, numrxjumps, num_sites)
-    hopping_rates = zeros(Float64, num_species, num_sites)
-    SpatialRates(reaction_rates,hopping_rates)
+function set_hop_rate_at_site!(hop_rates::HopRatesUnifNbr, site, species, rate) 
+    @inbounds old_rate = hop_rates.rates[species, site]
+    @inbounds hop_rates.rates[species, site] = rate
+    @inbounds hop_rates.sum_rates[site] += rate - old_rate
+    old_rate
+end
+
+function evalhoprate(hop_rates::HopRatesUnifNbr, u, species, site, num_nbs::Int) 
+    @inbounds u[species,site]*hop_rates.hopping_constants[species,site]*num_nbs
+end
+
+############## hopping rates of form L_{s,i,j} ################
+struct HopRatesGeneral{F} <: AbstractHopRates
+    hopping_constants::Vector{Matrix{F}} # hopping_constants[i][j,s] is the hopping constant at site i of species s to jth neighbor
+    rates::Vector{Matrix{F}} # rates[i][j,s] is the hopping rate at site i of species s to jth neighbor
+    sum_rates::Vector{F} # sum_rates[i] is the total hopping rate out of site i
 end
 
 """
-initializes SpatialRates with zero rates
+initializes HopRates with zero rates
 """
-function SpatialRates(ma_jumps, num_species, spatial_system::AbstractSpatialSystem)
-    num_sites = num_sites(spatial_system)
-    num_jumps = get_num_majumps(ma_jumps)
-    SpatialRates(num_jumps,num_species,num_sites)
+function HopRatesGeneral(hopping_constants::Vector{Matrix{F}}, transpose = true) where F <: Number
+    if transpose
+        hopping_constants = map(r -> r = collect(r'), hopping_constants) #transpose for better memory layout
+    end
+    rates = deepcopy(hopping_constants)
```

Comment:
For future PR we should optimize this to just assume they are all in the right layout from the user (to minimize copying internally).